### PR TITLE
Local API: Implement SABR for VODs

### LINKED ISSUE #6683: [Bug]: Regression:  0.23.0 selects/plays the wrong auto-dubbed audio track not the original like  0.22.1
### Guidelines

- [x] I have encountered this bug in the [latest release of FreeTube](https://github.com/FreeTubeApp/FreeTube/releases).
- [x] I have encountered this bug in the [official downloads of FreeTube](https://github.com/FreeTubeApp/FreeTube#official-downloads).
- [x] I have searched the issue tracker for [open](https://github.com/FreeTubeApp/FreeTube/issues?q=is%3Aopen+is%3Aissue) and [closed](https://github.com/FreeTubeApp/FreeTube/issues?q=is%3Aissue+is%3Aclosed) issues that are similar to the bug report I want to file, without success.
- [x] I have searched the [documentation](https://docs.freetubeapp.io/) for information that matches the description of the bug I want to file, without success.
- [x] This issue contains only one bug.

### Describe the bug

version  0.22.1  plays  this video ` 4rEnJO-fwic`  with the original  audio track
Player does not select the auto-dubbed English audio track. 

The new version starts playing  the English track, not the original German audio track ( this is consistent with local and invidious API)
THis happens every time then video begins to play.   I'm aware of issue https://github.com/FreeTubeApp/FreeTube/issues/4291 
but this looks like a regression.  

Changing the  'Locale Preference'  to 'German'  had no effect either.

0.22.1 does list the available audio tracks in different order  1. German  2. English 
0.23.0   shows English  then German. 

```
yt-dlp.exe -F https://youtu.be/4rEnJO-fwic
[youtube] Extracting URL: https://youtu.be/4rEnJO-fwic
[youtube] 4rEnJO-fwic: Downloading webpage
[youtube] 4rEnJO-fwic: Downloading tv client config
[youtube] 4rEnJO-fwic: Downloading player f3d47b5a
[youtube] 4rEnJO-fwic: Downloading tv player API JSON
[youtube] 4rEnJO-fwic: Downloading ios player API JSON
[youtube] 4rEnJO-fwic: Downloading m3u8 information
[info] Available formats for 4rEnJO-fwic:
ID    EXT   RESOLUTION FPS CH │    FILESIZE    TBR PROTO │ VCODEC           VBR ACODEC      ABR ASR MORE INFO
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
sb3   mhtml 48x27        0    │                    mhtml │ images                                   storyboard
sb2   mhtml 80x45        0    │                    mhtml │ images                                   storyboard
sb1   mhtml 160x90       0    │                    mhtml │ images                                   storyboard
sb0   mhtml 320x180      0    │                    mhtml │ images                                   storyboard
233-0 mp4   audio only        │                    m3u8  │ audio only           unknown             [en-US] American English - dubbed
234-0 mp4   audio only        │                    m3u8  │ audio only           unknown             [en-US] American English - dubbed
233-1 mp4   audio only        │                    m3u8  │ audio only           unknown             [de-DE] Deutsch (Deutschland) - original (default)
234-1 mp4   audio only        │                    m3u8  │ audio only           unknown             [de-DE] Deutsch (Deutschland) - original (default)
249-0 webm  audio only      2 │     5.04MiB    47k https │ audio only           opus        47k 48k [en-US] English (United States) (default), low, webm_dash
250-0 webm  audio only      2 │     6.10MiB    57k https │ audio only           opus        57k 48k [en-US] English (United States) (default), low, webm_dash
249-1 webm  audio only      2 │     5.34MiB    50k https │ audio only           opus        50k 48k [de-DE] German (Germany) original, low, webm_dash
250-1 webm  audio only      2 │     6.13MiB    57k https │ audio only           opus        57k 48k [de-DE] German (Germany) original, low, webm_dash
140-0 m4a   audio only      2 │    13.82MiB   129k https │ audio only           mp4a.40.2  129k 44k [en-US] English (United States) (default), medium, m4a_dash
140-1 m4a   audio only      2 │    13.82MiB   129k https │ audio only           mp4a.40.2  129k 44k [de-DE] German (Germany) original, medium, m4a_dash
251-0 webm  audio only      2 │    11.68MiB   109k https │ audio only           opus       109k 48k [en-US] English (United States) (default), medium, webm_dash
251-1 webm  audio only      2 │    11.38MiB   107k https │ audio only           opus       107k 48k [de-DE] German (Germany) original, medium, webm_dash
```

### Expected Behavior

play original audio track if possible. 
Works also in newpipe   and ReVanced  on Android.

### Issue Labels

inconsistent behavior

### FreeTube Version

0.23.0

### Operating System Version

Win10

### Installation Method

.zip / .7z

### Primary API used

Local API

### Last Known Working FreeTube Version (If Any)

0.22.1

### Additional Information

I noticed that this also happen with the browser (firefox) 

### Nightly Build

- [ ] I have encountered this bug in the latest [nightly build](https://docs.freetubeapp.io/development/nightly-builds).

### LINKED ISSUE #7119: [Bug]: Intermittent - "Audio formats are not available for this video"
### Guidelines

- [x] I have encountered this bug in the [latest release of FreeTube](https://github.com/FreeTubeApp/FreeTube/releases).
- [x] I have encountered this bug in the [official downloads of FreeTube](https://github.com/FreeTubeApp/FreeTube#official-downloads).
- [x] I have [searched the issue tracker for open and closed issues](https://github.com/FreeTubeApp/FreeTube/issues?q=is%3Aissue+sort%3Arelevance-desc) that are similar to the bug report I want to file, without success.
- [x] I have searched the [documentation](https://docs.freetubeapp.io/) for information that matches the description of the bug I want to file, without success.
- [x] This issue contains only one bug.

### Describe the bug

When opening a video from search results, or subscription page, rarely (perhaps 1/30 times) a message "Audio formats are not available for this video" is briefly displayed at the bottom of the program and the video will not play.
Going back and then re-opening the video seems to always resolve the issue, and the video then plays with no problem. 

### Expected Behavior

The video should play first time, as audio formats are available, as is evident of retrying the same video a second time and it working. 

### Issue Labels

inconsistent behavior

### FreeTube Version

v0.23.3 Beta

### Operating System Version

Microsoft Windows Version 10 22H2 (OS Build 19045:5608)

### Installation Method

.exe

### Primary API used

Local API

### Last Known Working FreeTube Version (If Any)

_No response_

### Additional Information

_No response_

### Nightly Build

- [ ] I have encountered this bug in the latest [nightly build](https://docs.freetubeapp.io/development/nightly-builds).

### LINKED ISSUE #7920: [Bug]: Specific video throws Error: This video is unavailable
### Guidelines

- [x] I have encountered this bug in the [latest release of FreeTube](https://github.com/FreeTubeApp/FreeTube/releases).
- [x] I have encountered this bug in the [official downloads of FreeTube](https://github.com/FreeTubeApp/FreeTube#official-downloads).
- [x] I have [searched the issue tracker for open and closed issues](https://github.com/FreeTubeApp/FreeTube/issues?q=is%3Aissue+sort%3Arelevance-desc) that are similar to the bug report I want to file, without success.
- [x] I have searched the [documentation](https://docs.freetubeapp.io/) for information that matches the description of the bug I want to file, without success.
- [x] This issue contains only one bug.
- [x] I have read and agree to follow the [rules](https://docs.freetubeapp.io/community/rules/).

### Describe the bug

The new update fixed the videos not loading bug... for two days. While it isn't showing up all of the time, given the trend of the last version, it will probably only get more common over time

### Expected Behavior

Image 

Image

### Issue Labels

content not loading

### FreeTube Version

v0.23.8 Beta

### Operating System Version

Windows 10

### Installation Method

.exe

### Primary API used

Local API

### Last Known Working FreeTube Version (If Any)

_No response_

### Additional Information

_No response_

### Nightly Build

- [ ] I have encountered this bug in the latest [nightly build](https://docs.freetubeapp.io/development/nightly-builds).

### PR DESCRIPTION
**Pull request by PikachuEXE**
## Pull Request Type

- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue

https://github.com/FreeTubeApp/FreeTube/pull/7145
https://github.com/FreeTubeApp/FreeTube/pull/7887
Closes #6683
Closes #7119 
Closes #7920

## Description

Based on https://github.com/FreeTubeApp/FreeTube/pull/7145 (Please read that PR first
Using https://github.com/LuanRT/googlevideo 4.0.4

- Implementation updated (different from #7145) to have playback session scoped stuff like SABR context & playbackCookie (SABR details)
- Player streaming timeout doubled (30s > 60s) to handle sometimes ~40MB response instead of ~10MB (at 2k quality, e.g. https://youtu.be/NmdXl7vXZ7g)
- New setting for enabling SABR in experimental section

Non-SABR
- Also updated the implementation for https://github.com/FreeTubeApp/FreeTube/pull/7887 to show pre-roll ads time in toast
- Toast now updated to be possible to show updated messages (hard coded to refresh per 1s currently)

## Screenshots 

Lazy maybe add later?

### A. SABR off
- Test pre-roll ads time toast
- Test various video playback (should be the same)

### B. SABR on
- Test various video playback (should be the same)

Find custom builds at https://github.com/PikachuEXE/FreeTube/actions/workflows/build.yml?query=branch%3Asabr-pr
Find cutting edge custom builds (I only use this branch for testing fixes myself, but can be outdated once I finished testing): https://github.com/PikachuEXE/FreeTube/actions/workflows/build.yml?query=branch%3Acustom-builds%2Fsabr

## Desktop

- **OS:**
- **OS Version:**
- **FreeTube version:**

## Additional context

Remaining tasks
- [x] Add/Update labels/tooltips/toasts & make them translatable (after text confirmed)
- [x] Handle infinite player reload - Allow max 3 player reloads for the same video before fallback to older DASH, if player loaded the reload doesn't count
- [x] Handle player reload during playback - Reload while keeping timestamp
- [x] Handle SABR playback delay in the middle - keep toast for now
- [x] Handle infinite SABR playback delay - reload player after (per req) backoff requested 5 times / cumulative delay going to be larger than timeout for request
- [x] Handle infinite next request policy received - reload player after (per req) next request policy received 100 times
- [x] Enable SABR by default
- [ ] Handle more edge cases if found

### Post PR
- Remove download feature (broken anyway)

### CODE REVIEW

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Didnt test yet but im very sure this will close the SABR related issues

#7119
#6683

**Comment by absidue** [comment]
I really hope it being an experimental setting that is disabled by default is just because you have some things left in the remaining tasks section...

**Comment by PikachuEXE** [comment]
I don't think it should be enabled by default for the first release
Otherwise there might be too many feedbacks to be handled
Whether it should remain in experimental section depends on how many edge case found and not handled well (Might be moved or not before next release

**Comment by absidue** [comment]
If we are going to keep it as a setting for the moment, it definitely should stay in the experimental settings and we should have a very clear plan for getting rid of it ASAP. The MWEB workaround is a workaround and is unlikely to keep working forever, so if it breaks while we still allow people to not use SABR we will have to keep finding new workarounds, which quite frankly I don't want to have to do when we know that SABR is the proper solution.

If we keep it a setting I would suggest enabling it by default but having the setting there as a temporary opt-out, that way we'll get more feedback but people can also disable it temporarily if they have issues, until you've decided it is stable enough.

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — COMMENTED** [review]
Both with and without SABR i get this error but that is on a server that is running through tor. If i connect to a normal server it does work

https://github.com/user-attachments/assets/a2e054d0-1b65-4b63-a8c2-9f8f5cda2b18

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Im not seeing the return of audio tracks -> #6683

**Comment by PikachuEXE** [comment]
Test fix not pushed this this PR yet: https://github.com/PikachuEXE/FreeTube/actions/runs/17907855602

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Seems to resolve https://github.com/FreeTubeApp/FreeTube/issues/7921 but we have to check with users after release

**Comment by ArthurKun21** [comment]
> Test fix not pushed this this PR yet: https://github.com/PikachuEXE/FreeTube/actions/runs/17907855602

Was using this and got hit with 404 earlier for a music

https://youtu.be/8tamOu8AwFE

[bundle-1758589261369.log](https://github.com/user-attachments/files/22482157/bundle-1758589261369.log)

It was only once so far.

**Comment by PikachuEXE** [comment]
@efb4f5ff-1298-471a-8973-3d47447115dc 
Audio track seems fixed (I don't have many videos to test)
https://github.com/FreeTubeApp/FreeTube/issues/7921 seems to be timeout so this PR raising the timeout for SABR responses (due to bigger response body sometimes) might or might not make it better
Auto reload currently only occurs when infinite retry detected (except redirect where none encountered yet)

@ArthurKun21 Let me know if it occurs more often. But could be YT testing new player which affects dev/current release too

**Comment by ArthurKun21** [comment]
From branch -> https://github.com/PikachuEXE/FreeTube/actions/workflows/build.yml?query=branch%3Asabr-pr

Build -> https://github.com/PikachuEXE/FreeTube/actions/runs/17933228357

https://youtu.be/ubfACjzvBOg

[bundle-1758599566888.log](https://github.com/user-attachments/files/22483818/bundle-1758599566888.log)

---

I go out of the window and then play the same video again. It played but like few seconds to the music it reloaded. Thought you might want the log for it

[bundle-1758599733091.log](https://github.com/user-attachments/files/22483842/bundle-1758599733091.log)

**Comment by PikachuEXE** [comment]
I got more reloads at the middle these few days
No idea why, adding debug stuff to my custom build to find out...

Update 1: HTTP 403 Not player JS related (still using `0004de42`) but no idea what causes it..
Update 2: Seems to be `Failed to extract signature decipher algorithm` all the time
Update 3: Testing new fix for audio track with less requests but not sure if that's the cause of 403 (so I enabled some logging and toasts - https://github.com/PikachuEXE/FreeTube/actions/runs/17939377956

**Comment by ArthurKun21** [comment]
https://github.com/PikachuEXE/FreeTube/actions/runs/17939377956

same music video -> https://youtu.be/ubfACjzvBOg

[bundle-1758619603841.log](https://github.com/user-attachments/files/22490632/bundle-1758619603841.log)

it still reloaded near the start

----

different music video https://youtu.be/EjlMPu5sEgw

[bundle-1758619987994.log](https://github.com/user-attachments/files/22490731/bundle-1758619987994.log)

same issue, reloaded near the start

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
- Should be sorted A-Z
- 2x Chinese is listed but i only see one on YT

VirtualBoxVM_Wbl6jvje6r

English should be listed above descriptive
VirtualBoxVM_rW00ywAChm

**Comment by absidue** [comment]
> English should be sorted above descriptive

We don't sort the audio tracks, that was the same before this PR. Lets keep this PR focused on implementing SABR, all unrelated bugs and feature requests should be left until later.

**Comment by PikachuEXE** [comment]
Implemented fallback and tested a few days
Fixes some of the issues but would still have issue (invalid poToken) sometimes (and I am not sure if YT issue or this PR issue)
You can test with https://github.com/PikachuEXE/FreeTube/actions/runs/18115890405

**Comment by PikachuEXE** [comment]
Latest build, just merged dev and with proper commits (i.e. only debug/workaround changes in a separate branch for this build, others all in this PR
Will make the new setting enabled by default next week if no major issues with playback (I expect you to test now if not started yet~

Update 1: forgot to put build link - https://github.com/PikachuEXE/FreeTube/actions/runs/18298073231

**Comment by PikachuEXE** [comment]
Before I forgot: It seems what player reload will be requested (when SABR enabled) once poToken expires
I got this once I forgot to close a window and try to resume (and it starts buffering
Not sure if works every time though

File: src/renderer/helpers/api/local.js [hunk 1]
```diff
@@ -393,40 +407,83 @@ export async function getLocalVideoInfo(id) {
...
+            const experiments = Object.fromEntries(new URLSearchParams(config.serializedExperimentFlags).entries())
+            Object.assign(allExperiments, experiments)
+          }
+        }
+      }
+      for (const [key, value] of Object.entries(allExperiments)) {
+        if (/po_token/.test(key)) {
+          console.warn(key, value)
+        }
+      }
```
Comment (absidue) [inline src/renderer/helpers/api/local.js]: Do we need to keep this debug logging?

File: src/renderer/helpers/api/local.js [hunk 2]
```diff
@@ -393,40 +407,83 @@ export async function getLocalVideoInfo(id) {
...
       console.error('Local API, poToken generation failed', error)
       throw error
     }
   }
 
-  let clientName = webInnertube.session.context.client.clientName
-
   const info = await webInnertube.getInfo(id, { po_token: contentPoToken })
+  const sabrCannotBeUsed = info.streaming_data?.server_abr_streaming_url == null ||
+    info.player_config?.media_common_config?.media_ustreamer_request_config?.video_playback_ustreamer_config == null
```
Comment (absidue) [inline src/renderer/helpers/api/local.js:468]: Was this a real scenario you encountered or is this just to be safe?

File: src/renderer/helpers/player/SabrSchemePlugin.js [hunk 3]
```diff
@@ -0,0 +1,828 @@
...
+ * @template T
+ * @param {import('googlevideo/shared-types').Part} part
+ * @param {{ decode: (data: Uint8Array) => T }} decoder
+ * @returns {T | undefined}
+ */
+function decodePart(part, decoder) {
+  if (!part.data.chunks.length) return undefined
+
+  try {
+    return decoder.decode(concatenateChunks(part.data.chunks))
```
Comment (absidue) [inline src/renderer/helpers/player/SabrSchemePlugin.js]: As concatenateChunks() always creates a new Uint8Array and we know that there are various parts that always fit into a single chunk (NextRequestPolicy, StreamProtectionStatus, SabrError, MediaHeader, ReloadPlaybackContext), it makes sense to skip the call to concatenateChunks if it isn't needed.
```suggestion
    const chunk = part.data.chunks.length === 1 ? part.data.chunks[0] : concatenateChunks(part.data.chunks)
    return decoder.decode(chunk)
```

File: src/renderer/helpers/api/local.js (hunk 1 above)
Comment (PikachuEXE) [reply src/renderer/helpers/api/local.js]: Forgot to remove, now removed

File: src/renderer/helpers/player/SabrSchemePlugin.js (hunk 3 above)
Comment (PikachuEXE) [reply src/renderer/helpers/player/SabrSchemePlugin.js]: Seems working, included

File: src/renderer/helpers/api/local.js (hunk 2 above)
Comment (PikachuEXE) [reply src/renderer/helpers/api/local.js:468]: This is an actual issue I encountered (and can be reproduced with both windows & macOS) so this is what fixed it
I have no idea it can be absent before that

Comment (absidue) [reply src/renderer/helpers/api/local.js:468]: Very odd.

Comment (PikachuEXE) [reply src/renderer/helpers/api/local.js:468]: It was https://youtu.be/JotWLTbDN5E but it became age restricted a day later
And I leave this here just to be safe at least

**Comment by absidue** [comment]
To fix the unknown type issues for the `googlevideo` imports could you please add these lines to the `compilerOptions` in the `jsconfig.json` file:
```
"module": "esnext",
"moduleResolution": "bundler",
```

Additionally the `SabrStreamState` type is referenced on lines 54 and 170 of the `SabrSchemePlugin` but is only defined on line 600, so Visual Studio Code treats it as if it doesn't exist, could you please move it earlier on in the file.

**Comment by PikachuEXE** [comment]
Updated

**Comment by PikachuEXE** [comment]
New build created: https://github.com/PikachuEXE/FreeTube/actions/runs/18454142459
With fix for data import/export (https://github.com/FreeTubeApp/FreeTube/pull/8106) (So I can test on windows)
And with https://github.com/FreeTubeApp/FreeTube/pull/8115 (Also removed hardcoded player ID)

**Comment by PikachuEXE** [comment]
New build: https://github.com/PikachuEXE/FreeTube/actions/runs/18482293990
No debug, with latest dev (with all fixes so no workaround needed)
SABR also now enabled by default
There should only be minor changes/fixes left if any (I hope

**Comment by PikachuEXE** [comment]
Maybe last build - https://github.com/PikachuEXE/FreeTube/actions/runs/18607715883
Fixed issues in https://github.com/PikachuEXE/FreeTube/issues/83 (maybe OS specific no idea)
Added a tweak to load less data for video/audio requests (no idea how to test if it really loads less data) taken from https://github.com/LuanRT/googlevideo/blob/googlevideo-v4.0.4/src/core/SabrStreamingAdapter.ts#L426

**Comment by ArthurKun21** [comment]
Hello, thanks for the work. Just wanna say for future reference.

The issue in https://github.com/PikachuEXE/FreeTube/issues/83 like the video getting stuck in the link https://youtu.be/8sJCPuweieA was present in both windows and mac machine. But the issue was already fixed with the succeeding commits

**Comment by pnaf** [comment]
I just wanted to add that the stuck player issue in that link was also happening on Linux (I was also encountering the issue at the end of https://youtu.be/vBynw9Isr28, for example). So it seems less likely that this is an OS-specific issue.

The new build with the workaround did fix that issue for me. Thanks!

File: src/renderer/helpers/player/SabrManifestParser.js [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+
+    /** @type {SabrManifest} */
+    const manifestData = JSON.parse(decodeURIComponent(uri.slice(uriPrefixLength)))
+
+    const presentationTimeline = new shaka.media.PresentationTimeline(0, 0, true)
+    presentationTimeline.setStatic(true)
+    presentationTimeline.setSegmentAvailabilityDuration(Infinity)
+    presentationTimeline.lockStartTime()
+    // Avoid player stuck at the end
+    // Reproducible on windows with some videos
```
Comment (pnaf) [inline src/renderer/helpers/player/SabrManifestParser.js]:
```suggestion
    // Reproducible on multiple OSes with some videos
```

Comment (PikachuEXE) [inline src/renderer/helpers/player/SabrManifestParser.js]:
```suggestion
    // Reproducible on some OSes with certain videos (possibly duration related)
```

**Comment by PikachuEXE** [comment]
Maybe last build 2 - https://github.com/PikachuEXE/FreeTube/actions/runs/18669891374
Fixed issues in https://github.com/PikachuEXE/FreeTube/issues/83 with better fix and re-tested by issue opener
Previous fix might not work in some cases (but can't really find an example yet) but this one should work for all
Details in https://github.com/PikachuEXE/FreeTube/issues/83#issuecomment-3420238657

**Comment by kommunarr** [comment]
Not sure why, but [this video](https://youtu.be/i25z70gPuAU) causes the SABR to get stuck in a loop refreshing the video constantly. Even after leaving it and trying to watch any other video, they also get stuck in a SABR loading loop until I terminate and restart FreeTube.

**Comment by PikachuEXE** [comment]
@kommunarr Sorry I can't reproduce SABR loop (reload loop?). It just starts playing right after backoff time
I have no idea if it's random or what (it's rare and doesn't happen again with the same video

**Comment by kommunarr** [comment]
Yes, an infinite reloading of the player. It still occurs consistently for me on that one video, although for unknown reasons, my other videos are still playing fine this session after encountering this error as opposed to last time. Here's what showed up in the console on the player reloading FWIW:

```
renderer.js:2 Player Error (category and code explainations here: https://shaka-player-demo.appspot.com/docs/api/shaka.util.Error.html)
Video ID: "i25z70gPuAU"
FreeTube player context: "loading dash/audio manifest and setting default quality in mounted"

Severity: CRITICAL (2)
Category: undefined (undefined)
Code: undefined (undefined)
Stack trace:
TypeError: Cannot read properties of null (reading 'Qa')
    at Of (renderer.js:2:460753)
    at Object.get [as isAudioOnly] (renderer.js:2:456898)
    at renderer.js:2:2399284
    at renderer.js:2:99774
    at renderer.js:2:101470
    at renderer.js:2:101711
```

**Comment by PikachuEXE** [comment]
If you see a toast about reading then it's handling SABR reload request (which is caused by unknown reason and hard to reproduce)
The error is raised due to reload (Running `player.isAudioOnly()` I guess and `player` is gone, so it's not causing it
I can try to make the error disappear

Update 1: Build with error fix (will still emit LOAD_INTERRUPTED but it's shaka behaviour): https://github.com/PikachuEXE/FreeTube/actions/runs/19053961955

**Comment by StevenNL2000** [comment]
@kommunarr Are you sure the reloading is infinite and not just long but finite? I've been using this branch and I've noticed that it sometimes has to reload 3 times for me before the video starts playing. My educated guess is that this corresponds to YouTube trying to show 3 preroll ads.

**Comment by PikachuEXE** [comment]
After 3 reloads I coded it to fallback to non SABR...

**Comment by SilentParrot** [comment]
I noticed that this [video](https://youtu.be/1XsS-dv4b1s) reloads a few times before ad preroll, with the following errors in console
```
renderer.js:2 [YOUTUBEJS][Parser]: ZU: Type mismatch, got ListItemView expected MenuServiceItem | MenuServiceItemDownload.
    at qO (renderer.js:2:1755273)
    at XO (renderer.js:2:1763857)
    at new QC (renderer.js:2:1549742)
    at XO (renderer.js:2:1764068)
    at ZO (renderer.js:2:1764352)
    at new tI (renderer.js:2:1550477)
    at XO (renderer.js:2:1764068)
    at new jN (renderer.js:2:1729160)
    at XO (renderer.js:2:1764068)
    at ZO (renderer.js:2:1764352)
renderer.js:2 [YOUTUBEJS][Parser]: ZU: Type mismatch, got ExpandableMetadata expected VideoDescriptionHeader | ExpandableVideoDescriptionBody | VideoDescriptionMusicSection | VideoDescriptionInfocardsSection | VideoDescriptionCourseSection | VideoDescriptionTranscriptSection | VideoDescriptionTranscriptSection | HorizontalCardList | ReelShelf | VideoAttributesSectionView | HowThisWasMadeSectionView.
    at qO (renderer.js:2:1755273)
    at XO (renderer.js:2:1763857)
    at ZO (renderer.js:2:1764352)
    at new tE (renderer.js:2:1570558)
    at XO (renderer.js:2:1764068)
    at new nE (renderer.js:2:1570795)
    at XO (renderer.js:2:1764068)
    at ZO (renderer.js:2:1764352)
    at YO (renderer.js:2:1762710)
    at new cU (renderer.js:2:1769382)```

**Comment by BitSynKed** [comment]
So this video here: https://youtu.be/FJGHove2WhA seems to consistently fail with SABR. I've tried the video over multiple days and restarting Freetube and it always seems to fail until it hits the fallback, where it then works perfectly fine. I'm on an M1 mac with macos 15.7.1
Here's the console output:

Screenshot 2025-11-09 at 8 28 12 PM

**Comment by PikachuEXE** [comment]
I can see some reload loop issue for random videos but after a few hours it's gone
Wondering if it's some YT protection/rate limiting/unknown logic
I can reproduce with your video but no idea how to fix yet

**Comment by PikachuEXE** [comment]
Custom build with workaround from https://github.com/LuanRT/googlevideo/issues/42 until official fix is available
https://github.com/PikachuEXE/FreeTube/actions/runs/19258907841

**Comment by PikachuEXE** [comment]
Included workaround since https://github.com/LuanRT/YouTube.js/pull/1092 is merged but not released
https://github.com/PikachuEXE/FreeTube/actions/runs/19450588003

Param to be used: `enableVoiceBoost` in https://github.com/LuanRT/googlevideo/blob/googlevideo-v4.0.4/docs/api/exports/protos/interfaces/ClientAbrState.md?plain=1#L131

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — COMMENTED** [review]
Created a test playlist with a mix of every type of video in it. I've received allot of reloads and backoff time varies from 4-5 sec to 14-15sec.

Just to be clear im not against the backoff time but when a video reloads and let me wait again and again it just doesnt feel right

 Im not sure if i feel comfortable with this annoyance level :/. I understand the choice to fallback to non-SABR a bit more now but like i said in previous conversations we shouldnt include non SABR options at all

https://github.com/user-attachments/assets/b377603d-7042-4c32-97c4-4a4a2a1d83b2

https://github.com/user-attachments/assets/2194ae69-e44a-46c6-b033-2e802c669d2b

https://github.com/user-attachments/assets/479ecb44-77b8-467e-a9a7-ebead3fa87e8

https://github.com/user-attachments/assets/f7f1400e-872c-4aa7-9faf-dd3a6bf98751

https://github.com/user-attachments/assets/29b8c4eb-1c5d-4647-9dbe-70c18238c18e

https://github.com/user-attachments/assets/1ad870bb-5ef2-426e-9e36-35e21a62deea

https://github.com/user-attachments/assets/a2d41b4f-d8af-4c6a-ab02-62a435d407c5

https://github.com/user-attachments/assets/38d62654-74b3-4cf7-bfda-e2805b987f33

https://github.com/user-attachments/assets/7164ef44-2167-46e7-ba72-99a454735bbf

https://github.com/user-attachments/assets/d966646d-422c-4bd0-a136-f090c07d9bed

**Comment by PikachuEXE** [comment]
[As I said above last month](https://github.com/FreeTubeApp/FreeTube/pull/8047#issuecomment-3544631251) the potential fix in upstream which allows me to implement the fix (no idea how to test before it's released
Is not released yet - https://github.com/LuanRT/YouTube.js/pull/1093
Maybe we can request for a sooner release?

**Comment by absidue** [comment]
Last time I checked webpack only validates imports against the `exports` field in a package's `package.json` file if you use a package import e.g. `import x from 'youtubei.js'` but if you instead specify the full relative path into the `node_modules` folder to that file, you can import any file you want. So you could import the `FormatXTags` protobuf decoder and detect the voice boost formats in FreeTube without waiting for a YouTube.js update.

**Comment by BitSynKed** [comment]
So I've noticed that occasionally when the video expires and auto reloads it ends up with multiple timers showing up. This leads to the video reloading and beginning to play normally, only for there to still be another timer that ends up reloading it again. There ends up being a bit of a jumble of different timers that forces the video to reload several times before it sorts itself out.

**Comment by PikachuEXE** [comment]
Maybe last build: https://github.com/PikachuEXE/FreeTube/actions/runs/21014343103
Please test

**Comment by MulverineX** [comment]
I gave this a shot and its working for me, haven't tested it that extensively though.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
working perfectly like reported earlier internally but when i set default formats to legacy in the settings it always plays dash

**Comment by PikachuEXE** [comment]
I got this and I have no idea how to fix yet
Update 1: No such issue on dev
image

**Comment by PikachuEXE** [comment]
Fixed
Legacy format requires it to be handled as if SABR disabled, and the ad time parsing also needs to be updated
Build: https://github.com/PikachuEXE/FreeTube/actions/runs/21084719641

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
New build: https://github.com/PikachuEXE/FreeTube/actions/runs/21094847809

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
remaining thing todo is to get rid of the toggle in experimental settings and the fallback to the workarounds

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]

**Review by absidue — APPROVED** [review]

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> remaining thing todo is to get rid of the toggle in experimental settings and the fallback to the workarounds

This will be addressed in a followup. So ready to merge!

**Review by ghost — APPROVED** [review]
Great work everyone ❤️
